src/rgw/rgw_data_sync.cc

```diff
@@ -5410,145 +5344,362 @@ int RGWSyncBucketCR::operate(const DoutPrefixProvider *dpp)
 }
 
 int RGWBucketPipeSyncStatusManager::run(const DoutPrefixProvider *dpp)
 {
-  list<RGWCoroutinesStack *> stacks;
+  struct item {
+    BucketSyncState state = BucketSyncState::Incremental;
+    uint64_t gen = 0;
+    ceph::real_time prev_progress;
+    ceph::real_time progress;
+    std::vector<rgw_bucket_sync_pair_info> pairs;
+    uint64_t num_shards = 0;
+  };
+  // Contains an entry for every peer that has not finished
+  // incremental sync of its last generation.
+  std::map<source*, item> work;
+
+  // Get the status of every peer. In incremental sync, this gives us
+  // the generation and shard count that is next needed to be run.
+  for (auto& s : sources) {
+    auto& i = work[&s];
+    auto status = read_status(dpp, s);
+    if (!status) {
+      ldpp_dout(dpp, -1) << "ERROR: Unable to fetch status for zone="
+			 << s.sc.source_zone << " error="
+			 << status.error() << dendl;
+      return status.error();
+    }
+    i.state = status->state;
+    i.gen = status->incremental_gen;
+    i.num_shards = status->shards_done_with_gen.size();
+  }
+
+  // Each peer is removed from the work map as it's completed.
+  while (!work.empty()) {
+    // Build up the work to do. This is separate so in the event of
+    // error we can reschedule the coroutines and rerun it.
+    for (auto& [s, i] : work) {
+      // Special case to handle full sync. Since full sync no longer
+      // uses shards and has no generations, we sync generation zero,
+      // shard zero.
+      if (i.state != BucketSyncState::Incremental) {
+	i.pairs.emplace_back();
+	auto& pair = i.pairs.back();
+	pair.source_bs.bucket = s->info.bucket;
+	pair.dest_bucket = s->dest;
+	pair.source_bs.shard_id = 0;
+      } else {
+	// Otherwise run sync on every shard.
+	i.pairs.resize(i.num_shards);
+	for (auto shard = 0u; shard < i.num_shards; ++shard) {
+	  auto& pair = i.pairs[shard];
+	  pair.source_bs.bucket = s->info.bucket;
+	  pair.dest_bucket = s->dest;
+	  pair.source_bs.shard_id = shard;
+	}
+      }
+    }
 
-  for (auto& mgr : source_mgrs) {
-    RGWCoroutinesStack *stack = new RGWCoroutinesStack(store->ctx(), &cr_mgr);
-    for (int i = 0; i < mgr->num_pipes(); ++i) {
-      stack->call(mgr->run_sync_cr(i));
+    // Now, build up and execute the coroutines, repeating until done,
+    // or until we max out our allowed retries without making progress.
+    int ret = 0;
+    static constexpr auto allowed_retries = 10u;
+    for (auto retries = 0u;
+	 (retries < allowed_retries) && (ret != 0);
+	 ++retries) {
+      list<RGWCoroutinesStack *> stacks;
+      for (auto& [s, i] : work) {
+	RGWCoroutinesStack *stack = new RGWCoroutinesStack(store->ctx(),
+							   cr_mgr.get());
+	for (auto& pair : i.pairs) {
+	  stack->call(sync_bucket_shard_cr(&s->sc, nullptr, pair,
+					   i.gen,
+					   sync_env.sync_tracer->root_node,
+					   &i.progress));
+	}
+	stacks.push_back(stack);
+      }
+      ret = cr_mgr->run(dpp, stacks);
+      if (ret == -ECANCELED) {
```

Comment:
from 'the outside', it's hard to check for specific errors like this. any number of the coroutines we spawned here could fail, but `run()` can only return a single error code - which i think is just the result of the last coroutine that finished. so `run()` may return success even if there were errors

to get a reliable error code out, you'd have to spawn a single coroutine that itself does the map/reduce (on 'the inside') to choose a return value. but i really wouldn't recommend trying to loop over `cr_mgr->run()` this way

i think the whole algorithm is better implemented as a composition of RGWCoroutines, without the synchronous pieces in the middle. then `RGWBucketPipeSyncStatusManager::run()` could spawn one top-level coroutine for each source, and ignore the return value of `cr_mgr->run()` entirely. that would just run until all of the sources finish, and assume that they've all logged whatever errors they needed to

maybe it breaks down to something like this?
```c++
namespace rgw::bucket_sync_run {
class SourceCR; // read sync status, loop over calls to GenCR
class GenCR; // loop over calls to ShardCR with limited concurrency
class ShardCR; // retry-loop over calls to sync_bucket_shard_cr()
```
i think it's important for the retry-loop to be at the bottom, per-shard level, so we only have to retry the single shard instead of all shards/sources

i also think we should be careful with per-shard concurrency, because some generations may have several thousand. if we spawn everything at once and they all send http requests to a single endpoint, it might start refusing connections or returning "503 SlowDown" errors due to rate limits. see the [`RGWShardCollectCR`](https://github.com/ceph/ceph/blob/f4d7090/src/rgw/rgw_sync.h#L515-L533) base class and the [`yield_spawn_window()`](https://github.com/ceph/ceph/blob/f4d7090/src/rgw/rgw_coroutine.h#L385-L389) macro for ways to control the concurrency

Reply:
Two questions:

(a) Since the current algorithm doesn't retry either, is it critical to reimplement things with retry using the current coroutines, or can that be added when we aren't under the RGWCoroutine regime?

(b) What's the correct behavior if a shard exhausts the allowed number of retries and/or hits ECANCELED? Do we abort the entire sync from that peer or do something else?